File: 0062-unique-paths/0062-unique-paths.py
class Solution(object):
    def uniquePaths(self, m, n):
        table = {(0, 0): 1}  # converted from if (r, c) == (0,0):

        def dp(r, c):  # (2, 6)
            # Don't have to set bounded range because no recursive calls are made to r-1 or c-1
            for i in range(r + 1):
                for j in range(c + 1):
                    if i == 0 or j == 0:  # Fill both the 1st row and col with 1
                        table[(i, j)] = table[(0, 0)]
                    if i > 0 and j > 0: # Combine the cell below and the cell to the right
                        table[(i, j)] = table[(i - 1, j)] + table[(i, j - 1)]
            return table[(r, c)]
        return dp(m - 1, n - 1)

File: 0062-unique-paths/0062-unique-paths.java
class Solution {
        public int uniquePaths(int m, int n) {

        int[][] memo = new int[m][n];
        for (int i = 0; i < m; i++) {
            Arrays.fill(memo[i], -1);
        }

        return dp(m - 1, n - 1, memo); // 종점을 리턴
    }

    private static int dp(int r, int c, int[][] memo) {
        // base condition
        // 시작점을 1로 설정
        if (r == 0 && c == 0) {
            memo[r][c] = 1;
            return memo[r][c];
        }
        int uniquePaths = 0;

        // 범위 설정 및 현재 memo에 값이 없을 때 함수 호출
        //! 범위 설정을 AND 조건으로 하면 false일 때 중괄호 부분을 실행하지 않기 때문에 따로따로 설정해줘야함
        if (memo[r][c] == -1) {
            if (r > 0) {
                uniquePaths += dp(r - 1, c, memo);
            }
            if (c > 0) {
                uniquePaths += dp(r, c - 1, memo);
            }
            memo[r][c] = uniquePaths;
        }

        return memo[r][c];

    }

}